app/src/main/java/com/duckduckgo/app/trackerdetection/TrackerDataLoader.kt

```diff
@@ -16,62 +16,56 @@
 
 package com.duckduckgo.app.trackerdetection
 
-import com.duckduckgo.app.trackerdetection.api.TrackerListService
+import android.support.annotation.WorkerThread
+import com.duckduckgo.app.trackerdetection.db.TrackerDataDao
 import com.duckduckgo.app.trackerdetection.model.TrackerNetworks
 import com.duckduckgo.app.trackerdetection.store.TrackerDataStore
-import io.reactivex.android.schedulers.AndroidSchedulers
-import io.reactivex.schedulers.Schedulers
 import timber.log.Timber
 import javax.inject.Inject
 
-
-class TrackerDataLoader @Inject constructor(private val trackerDetector: TrackerDetector,
-                                            private val trackerDataStore: TrackerDataStore,
-                                            private val trackerListService: TrackerListService,
-                                            private val networkTrackers: TrackerNetworks) {
+@WorkerThread
+class TrackerDataLoader @Inject constructor(
+        private val trackerDetector: TrackerDetector,
+        private val trackerDataStore: TrackerDataStore,
+        private val trackerDataDao: TrackerDataDao,
+        private val networkTrackers: TrackerNetworks) {
 
     fun loadData() {
+
+        Timber.i("Loading Tracker data")
+
+        // these are stored to disk, then fed to the C++ adblock module
         loadAdblockData(Client.ClientName.EASYLIST)
         loadAdblockData(Client.ClientName.EASYPRIVACY)
+
+        // stored in DB, then read into memory
         loadDisconnectData()
     }
 
-    private fun loadAdblockData(name: Client.ClientName) {
+    fun loadAdblockData(name: Client.ClientName) {
+        Timber.i("Looking for adblock tracker ${name.name} to load")
 
         if (trackerDataStore.hasData(name)) {
+            Timber.i("Found adblock tracker ${name.name}")
+
             val client = AdBlockClient(name)
             client.loadProcessedData(trackerDataStore.loadData(name))
+
+            // remove old client, then add the client again
+            trackerDetector.removeClient(client)
```

Comment:
Because this touches a JNI C++ component which allocates and deallocates memory the add and remove calls need to happen on the same thread across multiple syncs.

Reply:
Maybe you could annotate the methods to say which thread they are expected to run on, or handle that in the wrapper explicitly?

But am curious why the memory (de)allocation is not considered thread safe?  (I don't understand why it would matter which thread the memory is allocated and deallocated on as it's all in the same process space)

Reply:
I've looked into this and found some really old references to it being important but I can't find any definitive evidence. Happy to lean on your JNI experience and drop it for now and investigate later if we run into issues.

Reply:
How does this remove the old client?  The call here is removing the new client which the tracker won't know anything about it.  I can't see anything in the code which would indicate that the new client equates to the old client (e.g. hashCode/equals)

Would this be better off being a `removeClient(withName:)` type method?